config/trace.go

```diff
@@ -4,15 +4,174 @@
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	"go.opentelemetry.io/otel/sdk/resource"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
 	"go.opentelemetry.io/otel/trace"
 	"go.opentelemetry.io/otel/trace/noop"
 )
 
-func initTracerProvider(cfg configOptions) (trace.TracerProvider, shutdownFunc) {
+var errNoValidSpanExporter = errors.New("no valid span exporter")
+
+func initTracerProvider(cfg configOptions, res *resource.Resource) (trace.TracerProvider, shutdownFunc, error) {
 	if cfg.opentelemetryConfig.TracerProvider == nil {
-		return noop.NewTracerProvider(), noopShutdown
+		return noop.NewTracerProvider(), noopShutdown, nil
+	}
+	opts := []sdktrace.TracerProviderOption{
+		sdktrace.WithResource(res),
+	}
+	for _, processor := range cfg.opentelemetryConfig.TracerProvider.Processors {
+		sp, err := spanProcessor(cfg.ctx, processor)
+		if err != nil {
+			return noop.NewTracerProvider(), noopShutdown, err
+		}
+		opts = append(opts, sdktrace.WithSpanProcessor(sp))
+	}
+	tp := sdktrace.NewTracerProvider(opts...)
+	return tp, tp.Shutdown, nil
+}
+
+func spanExporter(ctx context.Context, exporter SpanExporter) (sdktrace.SpanExporter, error) {
+	if exporter.Console != nil {
+		return stdouttrace.New(
+			stdouttrace.WithPrettyPrint(),
+		)
+	}
+	if exporter.OTLP != nil {
+		switch exporter.OTLP.Protocol {
+		case protocolProtobufHTTP:
+			return initOTLPHTTPSpanExporter(ctx, exporter.OTLP)
+		case protocolProtobufGRPC:
+			return initOTLPgRPCSpanExporter(ctx, exporter.OTLP)
+		default:
+			return nil, fmt.Errorf("unsupported protocol %q", exporter.OTLP.Protocol)
+		}
+	}
+	return nil, errNoValidSpanExporter
+}
+
+func spanProcessor(ctx context.Context, processor SpanProcessor) (sdktrace.SpanProcessor, error) {
+	if processor.Batch != nil {
+		exp, err := spanExporter(ctx, processor.Batch.Exporter)
+		if err != nil {
+			return nil, err
+		}
+		return batchSpanProcessor(processor.Batch, exp)
+	}
+	if processor.Simple != nil {
+		exp, err := spanExporter(ctx, processor.Simple.Exporter)
+		if err != nil {
+			return nil, err
+		}
+		return sdktrace.NewSimpleSpanProcessor(exp), nil
+	}
+	return nil, fmt.Errorf("unsupported span processor type %v", processor)
+}
+
+func initOTLPgRPCSpanExporter(ctx context.Context, otlpConfig *OTLP) (sdktrace.SpanExporter, error) {
+	opts := []otlptracegrpc.Option{}
+
+	if len(otlpConfig.Endpoint) > 0 {
+		u, err := url.ParseRequestURI(normalizeEndpoint(otlpConfig.Endpoint))
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, otlptracegrpc.WithEndpoint(u.Host))
+		if u.Scheme == "http" {
+			opts = append(opts, otlptracegrpc.WithInsecure())
+		}
+	}
+
+	if otlpConfig.Compression != nil {
+		switch *otlpConfig.Compression {
+		case "gzip":
```

Comment:
Should we add consts like for `protocolProtobufHTTP`? Take notice it would be reused in `initOTLPHTTPSpanExporter`.

Reply:
added